Feat: Buffer api (#731)
dictu-lang/Dictu: docs/docs/standard-lib/buffer.md

---
layout: default
title: Buffer
nav_order: 4
parent: Standard Library
---

# Buffer
{: .no_toc }

## Table of contents
{: .no_toc .text-delta }

1. TOC
{:toc}

---

## Buffer
To make use of the Buffer module an import is required.

```js
import Buffer;
```

The maximum byte size for a buffer is `2147483647 - 1`

### Buffer.new() -> Result\<Buffer>

Returns a Result with a new buffer with the given size in bytes.

```cs
const buffer = Buffer.new(1024).unwrap();
print(buffer);
// <Buffer>
```

### Buffer.fromString(String) -> Result\<Buffer>

Returns a Result with a new buffer created from the given string.

```cs
const buffer = Buffer.fromString("Dictu!").unwrap();
print(buffer);
// <Buffer>
```

### Buffer.resize(Number) -> Result\<Number>

> Resizes the buffer to the given size, the argument needs to be greater then 0 or the function will return a error.

Reviewer: ```suggestion
Resizes the buffer to the given size. The argument needs to be greater than 0 or the function will return a error.
```

buffer.set(0, 68).unwrap();
print(buffer.string()); // "Dictu!"
```

### Buffer.subarray(Number: start -> Optional, Number: end -> Optional) -> Result\<Buffer>

Returns a new Buffer with the optional given start and end parameters.
* `start`: The start index within the buffer, default `0`
* `end`: The end index within the buffer(non inclusive), default: `buffer.len()`
```cs
const buffer = Buffer.fromString("Dictu!").unwrap();
const sub = buffer.subarray(0, buffer.len()-1).unwrap();
print(sub.string()); // "Dictu"
```

### Buffer.string() -> String

Returns a string representation of the buffer.
```cs
const buffer = Buffer.fromString("Dictu!").unwrap();
const str = buffer.string();
print(str); // "Dictu!"
```

### Buffer.len() -> Number

Returns the byte length of the buffer.
```cs
const buffer = Buffer.new(9).unwrap();
const len = buffer.len();
print(len); // 9
```

### Buffer.values() -> List

Returns a list with the integer values of the buffer.
```cs
const buffer = Buffer.fromString("Dictu!").unwrap();
const v = buffer.values();
print(v); // [68, 105, 99, 116, 117, 33]
```

### Buffer.writeString(Number, String) -> Result\<Nil>

> Sets a string into buffer given the starting index, if the string doesn't fit in the buffer a error is returned.

Reviewer: ```suggestion
Sets a string into buffer given the starting index. Iif the string doesn't fit in the buffer an error is returned.
```

Reviewer: I think there's an extra "i" in the fix.

Author: yeye i saw, https://github.com/dictu-lang/Dictu/pull/731/files#diff-e47adca1c96ac76e1106a054a4fcab1a04848982b6b34667870d13c04e2686e4R119

```cs
const buffer = Buffer.new(6).unwrap();
buffer.writeString(0, "Dictu!");
```

### Buffer.readString(Number: start -> Optional, Number: end -> Optional) -> Result\<String>

> Returns a String with the optional given start and end parameters, this works very similar to subarray.

Reviewer: ```suggestion
Returns a String with the optional given start and end parameters. This works very similar to subarray.
```

* `start`: The start index within the buffer, default `0`
* `end`: The end index within the buffer(non inclusive), default: `buffer.len()`
```cs
const buffer = Buffer.new(6).unwrap();
buffer.writeString(0, "Dictu!");
const sub = buffer.string(0, buffer.len()-1).unwrap();
print(sub) // "Dictu"
```

### Buffer.readUInt64LE(Number) -> Result\<Number>

Returns the u64(unsigned 8 byte integer in little endian) value given the starting index.

> Returns a result with the value or a error.

Reviewer: ```suggestion
Returns a result with the value or an error.
```

Reviewer: There's a few of these throughout the docs. I won't comment on all of them.
